src/main/java/org/sagebionetworks/workers/util/semaphore/WriteReadSemaphoreImpl.java

```diff
@@ -1,49 +1,75 @@
 package org.sagebionetworks.workers.util.semaphore;
 
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 import org.sagebionetworks.database.semaphore.CountingSemaphore;
 
 public class WriteReadSemaphoreImpl implements WriteReadSemaphore {
 
+	private static final Logger log = LogManager.getLogger(WriteReadSemaphoreImpl.class);
+
 	final CountingSemaphore countingSemaphore;
 	final int maxNumberOfReaders;
 
-
 	public WriteReadSemaphoreImpl(CountingSemaphore countingSemaphore, int maxNumberOfReaders) {
-		if(countingSemaphore == null) {
+		if (countingSemaphore == null) {
 			throw new IllegalArgumentException("CountingSemaphore cannot be null");
 		}
 		this.countingSemaphore = countingSemaphore;
 		this.maxNumberOfReaders = maxNumberOfReaders;
 	}
-	
+
 	@Override
-	public WriteLock getWriteLock(WriteLockRequest request) throws Exception {
-		if(request == null) {
+	public WriteLock getWriteLock(WriteLockRequest request) throws LockUnavilableException {
+		if (request == null) {
 			throw new IllegalArgumentException("Request cannot be null");
 		}
-		WriteLockImpl lock = new WriteLockImpl(countingSemaphore, request);
+		WriteLockImpl lock = createWriteLock(request);
 		try {
 			lock.attemptToAcquireLock();
 			return lock;
 		} catch (Exception e) {

```

Comment:
Might be possible to use a try finally block:
try {
...
} catch (LockUnavilableException e) {
  throw e;
} catch (Exception e) {
  throw new RuntimeException(e);
} finally {
  try {
    lock.close();
  } catch (Exception e) {
    log.error(e);
  }
}